src/helper/DirectHelper.sol

```diff
@@ -0,0 +1,120 @@
+    function scaledBalanceOf(address) external view returns (uint256);
+    function decimals() external view returns (uint8);
+}
+
+interface InterestRateStrategyLike {
+    function calculateInterestRates(
+        address reserve,
+        uint256 availableLiquidity,
+        uint256 totalStableDebt,
+        uint256 totalVariableDebt,
+        uint256 averageStableBorrowRate,
+        uint256 reserveFactor
+    ) external view returns (
+        uint256,
+        uint256,
+        uint256
+    );
+}
+
+// Helper functions for keeper bots
+contract DirectHelper {
+
+    // --- Math ---
+    function _mul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require(y == 0 || (z = x * y) / y == x, "DssDirectDepositAaveDai/overflow");
+    }
+    uint256 constant RAY  = 10 ** 27;
+    function _rdiv(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = _mul(x, RAY) / y;
+    }
+
+    // Use this to determine whether exec() should be called based on your interest rate deviation threshold
+    // This assumes normal operation, culled / global shutdown should be handled externally
+    // Also assumes no liquidity issues
+    function shouldExec(
+        address _direct,
+        uint256 interestRateTolerance
+    ) external view returns (bool) {
+        // IMPORTANT: this function assumes Vat rate of this ilk will always be == 1 * RAY (no fees).
+        // That's why this module converts normalized debt (art) to Vat DAI generated with a simple RAY multiplication or division
+        // This module will have an unintended behaviour if rate is changed to some other value.
+
+        DirectLike direct = DirectLike(_direct);
+        VatLike vat = VatLike(direct.vat());
+        TokenLike dai = TokenLike(direct.dai());
+        address adai = direct.adai();
+        TokenLike stableDebt = TokenLike(direct.stableDebt());
+        TokenLike variableDebt = TokenLike(direct.variableDebt());
+        bytes32 ilk = direct.ilk();
+        InterestRateStrategyLike interestStrategy = InterestRateStrategyLike(direct.interestStrategy());
+
+        (uint256 daiDebt,) = vat.urns(ilk, address(direct));
+        uint256 _bar = direct.bar();
+        if (_bar == 0) {
+            return daiDebt > 0;     // Always attempt to close out if we have debt remaining
+        }
+
+        (,, uint256 currVarBorrow) = interestStrategy.calculateInterestRates(
+            adai,
+            dai.balanceOf(adai),
+            stableDebt.totalSupply(),
+            variableDebt.totalSupply(),
+            0,
+            0
+        );
+
+        uint256 deviation = _rdiv(currVarBorrow, _bar);
+        if (deviation < RAY) {
+            // Unwind case
+            return daiDebt > 0 && (RAY - deviation) > interestRateTolerance;
+        } else if (deviation > RAY) {
+            // Wind case
+            (uint256 Art,,, uint256 line,) = vat.ilks(ilk);
+            return Art*RAY < line && (deviation - RAY) > interestRateTolerance;
```

Comment:
should this be `>= interestRateTolerance`?

Reply:
This does not calculate whether or not there is room in the debt ceiling to succeed in the `wind` so if we are 1 wei below `line` and outside our tolerance the bot would call `exec` but would revert, right? Do you think it is worth calling `calculateTargetSupply` to see if there is room in the `line` to wind?

Reply:
Yeah I thought about this case, and I think it will exactly max out whenever it can, so it's really unlikely to be within some small delta of the debt ceiling.

Anyways, this is not a high stakes contract to get perfect. Worst case we burn a bit of gas and redeploy.

Reply:
oh you're right it takes either the amount to add to change the rate or the delta to fill up the `line` 👍🏻